### PR TITLE
CPLAT-15554 Fix render swallowing errors on component mount

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
When `render`ing a component that throws on mount, React warnings that something went wrong are printed, but the actual error itself is missing.

This is because `render` uses `recordConsoleLogs`, which catches and swallows errors.

## Changes
- Don't catch errors thrown during `render`, or any other place we're recording console logs.
   - This involved slightly refactoring `recordConsoleLogs` into `spyOnConsoleLogs`, which allows handling logs even if the callback function throws
   - Add regression test along for functional-style test that related logs are forwarded to the console when a component throws on mount
- Update console log --> `print` forwarding to print logs as they come through, instead of at the end of a callback, using `printConsoleLog`
- Where possible, update tests to verify the `print`ed values, not just the recorded console log values, to more functionally test that the logs we expect make their way to the consumer's terminal when running tests.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author:
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

File: lib/src/util/console_log_utils.dart [hunk 1]
```diff
@@ -14,57 +14,89 @@
...
+    stopSpying();
+  }
+}
+
+/// Starts spying on concole logs, calling [onLog] for each log that occurs until the
+/// returned function (`stopSpying`) is called.
 ///
-/// If any errors are thrown during the callback, e.g. during a render that expects
-/// props that are not valid, the errors will be caught to allow the test to complete.
-List<String> recordConsoleLogs(
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:35]: `recordConsoleLogs` still exists, but was moved to `test/unit/util/prints_and_logs_recording.dart`

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by aaronlademann-wf — DISMISSED** [review]
Looks good. Just one super minor thing.

File: test/unit/console_log_utils_test.dart [hunk 2]
```diff
@@ -38,7 +41,96 @@ bool _assertsEnabled() {
...
+      test('does not swallow errors that occur in the callback', () {
+        expect(() {
+          spyOnConsoleLogs(() => throw ExceptionForTesting(), onLog: (_) {});
+        }, throwsA(isA<ExceptionForTesting>()));
+      });
+
+      test('calls onLog in the same zone as the spyOnConsoleLogs call', () {
+        final logCallZones = <Zone>[];
+        final testZone = Zone.current.fork()
+          ..run(() {
```
Comment (joebingham-wk) [inline test/unit/console_log_utils_test.dart:89]: oh_my
TIL you could do this haha

**Review by joebingham-wk — DISMISSED** [review]
~+10~

Edit: was a +10, but misread CI. Looks like tests and formatting are having a hard time

File: test/unit/console_log_utils_test.dart [hunk 3]
```diff
@@ -38,7 +41,96 @@ bool _assertsEnabled() {
...
+
+        window.console.log('foo');
+        window.console.log('bar');
+        expect(logCalls, ['foo', 'bar']);
+
+        stopSpying();
+
+        logCalls.clear();
+        window.console.log('baz');
+        expect(logCalls, []);
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/unit/console_log_utils_test.dart:58]: Yeah, I kept these as `[]` for consistency with the expectations that they contained certain items. In general though I don't think that `isEmpty` is necessarily preferable to expecting it to equal an empty list.

**Review by aaronlademann-wf — DISMISSED** [review]
+1 refresh

**Review by joebingham-wk — APPROVED** [review]
+10 refresh

I double checked with Greg and learned that this CI is expected to look as it does now 👍

**Review by joebingham-wk — APPROVED** [review]
+10

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**